extensions/ql-vscode/src/data-extensions-editor/model-details/model-details-data-provider.ts

```diff
@@ -34,16 +34,27 @@ export class ModelDetailsDataProvider
     return this.onDidChangeTreeDataEmitter.event;
   }
 
+  /**
+   * Update the data displayed in the tree view.
+   *
+   * Will only trigger an update if the data has changed. This relies on
+   * object identity, so be sure to not mutate the data passed to this
+   * method and instead always pass new objects/arrays.
```

Comment:
Would it make sense to enforce this in the callers by always using a `ReadonlyArray`?

Reply:
I don't think it'll be simple, so I'm not sure if the benefit is worth the inconvenience.

I gave it a quick try and the scope quickly grows and we need to mark more and more arrays as readonly. It could be helpful still, but I think it's a large change of direction so I don't feel comfortable making it here. Investigating this further in another issue could be good, and we could consider making other types like `ExternalApiUsage` immutable as well.

My understanding is that in typescript you can cast a mutable array to an immutable array, but not the other way around. This means if we just specify `setState` to take a readonly array it doesn't gain us anything because the caller is still holding a mutable reference. Therefore we have to declare it as readonly as early in the process as possible, but this ends up infecting more and more of the codebase that isn't directly related to the details panel.

Reply:
Thanks for exploring this path! It makes sense that there are too many places where we would need to make updates, and it probably doesn't really make a big difference in type safety anyway. It would just increase the number of places where we need to specify types.